integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/common/BenchUtils.scala

```diff
@@ -0,0 +1,399 @@
+      val hotRuns = queryTimes.drop(1)
+      val numHotRuns = hotRuns.length
+      println(s"Best of $numHotRuns hot run(s): ${hotRuns.min} msec.")
+      println(s"Worst of $numHotRuns hot run(s): ${hotRuns.max} msec.")
+      println(s"Average of $numHotRuns hot run(s): " +
+          s"${hotRuns.sum.toDouble/numHotRuns} msec.")
+    }
+
+    // write results to file
+    val filename = s"$filenameStub-${queryStartTime.toEpochMilli}.json"
+    println(s"Saving benchmark report to $filename")
+
+    // try not to leak secrets
+    val redacted = Seq("TOKEN", "SECRET", "PASSWORD")
+    val envVars: Map[String, String] = sys.env
+        .filterNot(entry => redacted.exists(entry._1.toUpperCase.contains))
+
+    val testConfiguration = TestConfiguration(
+      gcBetweenRuns
+    )
+
+    val environment = Environment(
+      envVars,
+      sparkConf = df.sparkSession.conf.getAll,
+      getSparkVersion)
+
+    val queryPlan = QueryPlan(
+      df.queryExecution.logical.toString(),
+      df.queryExecution.executedPlan.toString()
+    )
+
+    val report = resultsAction match {
+      case Collect() => BenchmarkReport(
+        filename,
+        queryStartTime.toEpochMilli,
+        environment,
+        testConfiguration,
+        "collect",
+        Map.empty,
+        queryDescription,
+        queryPlan,
+        queryTimes)
+
+      case w: WriteCsv => BenchmarkReport(
+        filename,
+        queryStartTime.toEpochMilli,
+        environment,
+        testConfiguration,
+        "csv",
+        w.writeOptions,
+        queryDescription,
+        queryPlan,
+        queryTimes)
+
+      case w: WriteParquet => BenchmarkReport(
+        filename,
+        queryStartTime.toEpochMilli,
+        environment,
+        testConfiguration,
+        "parquet",
+        w.writeOptions,
+        queryDescription,
+        queryPlan,
+        queryTimes)
+    }
+
+    writeReport(report, filename)
+  }
+
+  def readReport(file: File): BenchmarkReport = {
+    implicit val formats = DefaultFormats
+    val json = parse(file)
+    json.extract[BenchmarkReport]
+  }
+
+  def writeReport(report: BenchmarkReport, filename: String): Unit = {
+    implicit val formats = DefaultFormats
+    val os = new FileOutputStream(filename)
+    os.write(writePretty(report).getBytes)
+    os.close()
+  }
+
+  def getSparkVersion: String = {
+    // hack for databricks, try to find something more reliable?
+    if (SPARK_BUILD_USER.equals("Databricks")) {
+      SPARK_VERSION + "-databricks"
+    } else {
+      SPARK_VERSION
+    }
+  }
+
+  /**
+   * Perform a diff of the results collected from two DataFrames, allowing for differences in
+   * precision.
+   *
+   * This is only suitable for data sets that can fit in the driver's memory.
+   */
+  def compareResults(df1: DataFrame, df2: DataFrame, epsilon: Double = 0.00001): Unit = {
+
+    val result1: Seq[Seq[Any]] = collectAndSort(df1)
+    val result2: Seq[Seq[Any]] = collectAndSort(df2)
```

Comment:
What if ordering of the results is a significant part of the query output?  There are cases where we need to preserve order when comparing.  For example, if I run a query where I'm collecting the top 100, having the order be off but the same 100 elements will declare victory when it should not.

Minimally this needs to account for unordered vs. ordered compare.